File: crates/sc-search/src/varint.rs
//! LEB128-style unsigned varints. Posting lists are delta+varint encoded
//! (§4.1); block payloads use the same encoding for lengths.

/// Append `v` to `out`.
pub fn put(out: &mut Vec<u8>, mut v: u64) {
    loop {
        let b = (v & 0x7f) as u8;
        v >>= 7;
        if v == 0 {
            out.push(b);
            return;
        }
        out.push(b | 0x80);
    }
}

/// Read a varint at `*pos`, advancing it. `None` on truncation or overlong
/// encodings.
pub fn get(buf: &[u8], pos: &mut usize) -> Option<u64> {
    let mut v: u64 = 0;
    let mut shift = 0u32;
    loop {
        let b = *buf.get(*pos)?;
        *pos += 1;
        if shift >= 64 {
            return None;
        }
        v |= ((b & 0x7f) as u64) << shift;
        if b & 0x80 == 0 {
            return Some(v);
        }
        shift += 7;
    }
}

/// Encoded width of `v`, in bytes. Used by the size estimator (§6.3).
pub fn len(v: u64) -> usize {
    let mut n = 1;
    let mut v = v >> 7;
    while v != 0 {
        n += 1;
        v >>= 7;
    }
    n
}

/// Delta+varint encode a strictly ascending list of block ids.
pub fn encode_ascending(out: &mut Vec<u8>, ids: &[u32]) {
    let mut prev = 0u64;
    for (i, &id) in ids.iter().enumerate() {
        let id = id as u64;
        if i == 0 {
            put(out, id);
        } else {
            put(out, id - prev);
        }
        prev = id;
    }
}

/// Inverse of [`encode_ascending`].
pub fn decode_ascending(buf: &[u8]) -> Option<Vec<u32>> {
    let mut out = Vec::new();
    let mut pos = 0usize;
    let mut prev = 0u64;
    while pos < buf.len() {
        let d = get(buf, &mut pos)?;
        let v = if out.is_empty() { d } else { prev + d };
        if v > u32::MAX as u64 {
            return None;
        }
        out.push(v as u32);
        prev = v;
    }
    Some(out)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn roundtrip() {
        for v in [0u64, 1, 127, 128, 300, 16383, 16384, u32::MAX as u64, u64::MAX] {
            let mut buf = Vec::new();
            put(&mut buf, v);
            assert_eq!(buf.len(), len(v), "len mismatch for {v}");
            let mut p = 0;
            assert_eq!(get(&buf, &mut p), Some(v));
            assert_eq!(p, buf.len());
        }
    }

    #[test]
    fn ascending_roundtrip() {
        let ids: Vec<u32> = vec![0, 1, 5, 400, 401, 100_000];
        let mut buf = Vec::new();
        encode_ascending(&mut buf, &ids);
        assert_eq!(decode_ascending(&buf).unwrap(), ids);
    }

    #[test]
    fn truncated_is_none() {
        let mut buf = Vec::new();
        put(&mut buf, 300);
        buf.pop();
        let mut p = 0;
        assert_eq!(get(&buf, &mut p), None);
    }
}

File: go/internal/search/varint.go
// Package search implements the tiered filename search: the parallel walk, the
// optional trigram index, and the estimator that chooses between them.
//
// The on-disk format is not changing, which is what makes this a translation
// with a golden-file check rather than a reimplementation.
package search

import (
	"errors"
	"math"
)

// LEB128-style unsigned varints. Posting lists are delta plus varint encoded,
// and block payloads use the same encoding for their lengths.

// ErrVarint is a truncated or overlong encoding.
var ErrVarint = errors.New("search: a malformed varint")

// PutVarint appends v to out.
func PutVarint(out []byte, v uint64) []byte {
	for {
		b := byte(v & 0x7f)
		v >>= 7
		if v == 0 {
			return append(out, b)
		}
		out = append(out, b|0x80)
	}
}

// Varint reads a varint at pos, returning the value and the position after it.
//
// Only the canonical encoding of a value is accepted. A continuation byte that
// would shift past 64 bits is refused, and so is a final byte contributing no
// bits: both are second spellings of a number the encoder never writes, and a
// decoder that took them could not re-encode what it just read. Refusing means
// a corrupt segment is detected here rather than turned into plausible hits.
func Varint(buf []byte, pos int) (uint64, int, error) {
	var v uint64
	var shift uint
	for {
		if pos < 0 || pos >= len(buf) {
			return 0, 0, ErrVarint
		}
		b := buf[pos]
		pos++
		if shift >= 64 {
			return 0, 0, ErrVarint
		}
		// The tenth byte can only carry the single bit that reaches bit 63.
		// Anything above it would be dropped by the shift, so a value that set
		// one is a corrupt encoding rather than a large number.
		if shift == 63 && b&0x7f > 1 {
			return 0, 0, ErrVarint
		}
		v |= uint64(b&0x7f) << shift
		if b&0x80 == 0 {
			// A final byte contributing no bits is the overlong form of a
			// shorter encoding.
			if shift > 0 && b == 0 {
				return 0, 0, ErrVarint
			}
			return v, pos, nil
		}
		shift += 7
	}
}

// VarintLen is the encoded width of v in bytes, which the size estimator needs
// without encoding anything.
func VarintLen(v uint64) int {
	n := 1
	for v >>= 7; v != 0; v >>= 7 {
		n++
	}
	return n
}

// PutAscending delta-encodes a strictly ascending list of block ids.
//
// The first id is absolute and the rest are gaps, which is what makes a dense
// posting list mostly one-byte values.
func PutAscending(out []byte, ids []uint32) []byte {
	var prev uint64
	for i, id := range ids {
		v := uint64(id)
		if i == 0 {
			out = PutVarint(out, v)
		} else {
			out = PutVarint(out, v-prev)
		}
		prev = v
	}
	return out
}

// Ascending is the inverse of PutAscending.
func Ascending(buf []byte) ([]uint32, error) {
	var out []uint32
	var prev uint64
	pos := 0
	for pos < len(buf) {
		d, next, err := Varint(buf, pos)
		if err != nil {
			return nil, err
		}
		pos = next
		v := d
		if len(out) > 0 {
			v = prev + d
		}
		if v > math.MaxUint32 {
			return nil, ErrVarint
		}
		out = append(out, uint32(v))
		prev = v
	}
	return out, nil
}
